src/sage/combinat/integer_vector.py

```diff
@@ -968,6 +1055,78 @@ def __contains__(self, x):
             return False
         return len(x) == self.k
 
+    def rank(self, x):
+        """
+        Return the rank of a given element.
+
+        INPUT:
+
+        - ``x`` -- a list with ``len(x) == k``
+
+        EXAMPLES::
+
+            sage: IntegerVectors(k=5).rank([0,0,0,0,0])
+            0
+            sage: IntegerVectors(k=5).rank([1,1,0,0,0])
+            7
+        """
+        if len(x) != self.k:
+            raise ValueError("argument is not a member of IntegerVectors({},{})".format(None, self.k))
+
+        n, k = sum(x), self.k
+
+        r = sum(binomial(k + i - 1, k - 1) for i in range(n))
```

Comment:
It may be useful to notice that
` sum(binomial(k + i - 1, k - 1) == binomial(n+k-1, k)`
avoiding summation (please check yourself before using).

Currently it seems that `IntegerVectors(k=1)[n]` takes time roughly quadratic to `n`:
```
sage: V=IntegerVectors(k=1)
....: %time V[1000]
CPU times: user 791 ms, sys: 0 ns, total: 791 ms
Wall time: 791 ms
[1000]
sage: V=IntegerVectors(k=1)
....: %time V[2000]
CPU times: user 3.13 s, sys: 0 ns, total: 3.13 s
Wall time: 3.13 s
[2000]
sage: V=IntegerVectors(k=1)
....: %time V[4000]
CPU times: user 10.9 s, sys: 0 ns, total: 10.9 s
Wall time: 10.9 s
[4000]
```

How does it fare if you replace the summation with a single binomial?